chat-android/src/main/java/com/ably/chat/TaskScheduler.kt

```diff
@@ -0,0 +1,69 @@
+package com.ably.chat
+
+import java.util.PriorityQueue
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.async
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.launch
+
+private class Task(
+    private val priority: Int,
+    val coroutineBlock: suspend CoroutineScope.() -> Any,
+    val result: TaskResult<Any>)
+    : Comparable<Task> {
+        override fun compareTo(other: Task): Int {
+            return this.priority.compareTo(other.priority)
+        }
+
+        suspend fun setResult(res: Result<Any>) {
+            result.channel.send(res)
+        }
+}
+
+class TaskResult<T> {
```

Comment:
I believe it's better to use framework's implementation for deferred value - [CompletableDeferred](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-completable-deferred/)